RFC-0008-plan-checker.md

```diff
@@ -0,0 +1,137 @@
+# **RFC0 for Presto**
+
+See [CONTRIBUTING.md](CONTRIBUTING.md) for instructions on creating your RFC and the process surrounding it.
+
+## [Title]
+
+Proposers
+
+* Tim Meehan
+* Bryan Cutler
+* Rebecca Schlussel
+
+## [Related Issues]
+
+* RFC-0003
+
+## Summary
+
+Add a new SPI to integrate a custom plan checker, and add a plugin to use the Presto sidecar to check if a Presto plan can be
+successfully translated into a Velox plan.
+
+## Background
+
+The optimizer makes decisions in part based on the capabilities of the underlying evaluation engine.  With the Presto evaluation
+engine being migrated to Velox while concurrently supporting the Presto Java evaluation engine, there's now differences between
+what is supported between both evaluation engines.  Because of these underlying differences, the optimizer may generate plans that
+can't be executed in C++ clusters, or likewise, a Presto Java cluster could be misconfigured to generate plans that only work with
+C++ clusters.
+
+If a plan is generated that can't be executed in the cluster, the query will fail with an error message during the query execution
+phase.  This is a poor user experience, as the user has to wait for the query to fail before they can take corrective action, and
+this might have occurred after lengthy queueing.  Additionally, this failure would occur after worker resources have already been
+allocated to the query, which is wasteful.  This RFC proposes a plan checker that can be run before the query is executed to ensure
+a quick validation of the plan.
+
+### [Optional] Goals
+
+* Provide a mechanism to validate Presto to Velox plan conversion during planning phase
+* Add an SPI to allow custom validators to be added to suit individual business needs
+
+### [Optional] Non-goals
+
+* Ensure all Velox plans are executable in a Velox cluster--many checks are done at runtime and may not be caught by the plan checker
+
+## Proposed Implementation
+
+The Presto runtime centralizes plan validation logic into the `PlanChecker` class.  There exist three phases to this class: 
+
+* `validateIntermediatePlan`
+* `validateFinalPlan`
+* `validateFragmentedPlan`
+
+It is generally useful to allow this class to be configured with an SPI that allows for custom plan validation logic to be added.
+For example, a business may decide that a certain type is not allowed, or add a check to ensure that plans that are overly
+complicated are killed.
+
+An SPI will be added that will add more checks to the `PlanChecker` class.  The SPI will contain a field indicating which phase
+of the plan checker to be added, and a validator that will be run during that phase.  The SPI will be added to the `PlanChecker`
```

Comment:
or rather, i think custom plan checkers are useful, but specifically velox plan translation could be independently useful and possibly faster to implement as an extension of explain.

Reply:
I got this info from Sergey when I asked about it:

> Velox has "struct PlanFragment", which is what you are after. Adding toString() or toJson() method to it might work. PlanFragment has  std::shared_ptr<const core::PlanNode> planNode; which has rich toString() already.


I'm okay with both approaches, but let's make the post request be to a path specific for validation, so in the future we could extend to be able to get the full serialized plan if we want to.

Reply:
I also see that PlanFragment is already serializable as a string.

I may not quite understand your original suggestion.  We are using a plan checker to provide validation of the plan prior to query execution, so we don't wait until execution to discover that the generated plan won't be able to execute.

This is tied to the lifecycle of the query.  It sounds like what you're suggesting is, add a new explain plan type that attempts to translate the plan to a Velox plan, and throw if it's not possible?  If so, I think we can do something along those lines too.

1) The new explain type could use the sidecar and SPI created in this proposal to accomplish this.
2) The output may contain a combination of Velox and Presto plans, however, since some fragments are coordinator only.

I think it's a good idea and we can include that in this proposal, as it would also allow for a non-intrusive way to determine if the plan could be executed in Presto C++.  However, I wanted to make sure this is what you're proposing.

Reply:
yeah, exactly.  The benefit is a bit tangential to the goal of query validation, as with the plan checker approach we can still validate the query by running explain(type distributed) with the plan checker enabled, but since we're doing all the work to translate the query plan, we may as well add the ability to print it.

Reply:
I have added a section on the extended plan format.